rclpy/src/rclpy_common/src/common.c

```diff
@@ -338,3 +338,96 @@ rclpy_convert_to_py(void * message, PyObject * pyclass)
   }
   return convert(message);
 }
+
+PyObject *
+rclpy_convert_to_py_topic_info_list(const rmw_topic_info_array_t * info_array)
+{
+  if (!info_array) {
+    return NULL;
+  }
+
+  PyObject * py_info_array = PyList_New(info_array->count);
+  if (!py_info_array) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < info_array->count; ++i) {
+    rmw_topic_info_t topic_info = info_array->info_array[i];
+    const rmw_qos_profile_t * qos_profile = topic_info.qos_profile;
+    PyObject * py_qos_profile = rclpy_common_convert_to_qos_dict(qos_profile);
+    if (!py_qos_profile) {
+      Py_DECREF(py_info_array);
```

Comment:
To improve error handling, I would add a function `rclpy_convert_to_py_topic_info(const rmw_topic_info_t * info)`:

```c
rclpy_convert_to_py_topic_info(const rmw_topic_info_t * topic_info)
{
    PyObject * py_qos_profile = NULL;
    PyObject * py_node_name = NULL;
    PyObject * py_node_namespace = NULL;
    PyObject * py_topic_type = NULL;
    PyObject * py_gid = NULL;
    PyObject * py_topic_info_dict = NULL;

    py_qos_profile = rclpy_common_convert_to_qos_dict(topic_info.qos_profile);
    if (!py_qos_profile) {
      goto fail;
    }

    py_node_name = PyUnicode_FromString(topic_info.node_name);
    if (!py_node_name) {
      goto fail;
    }

    py_node_namespace = PyUnicode_FromString(topic_info.node_namespace);
    if (!py_node_namespace) {
      goto fail;
    }

    py_topic_type = PyUnicode_FromString(topic_info.topic_type);
    if (!py_topic_type) {
      goto fail;
    }

    py_gid = PyUnicode_FromString(topic_info.gid);
    if (!py_gid) {
      goto fail;
    }

    // Create dictionary that represents rmw_topic_info_t
    py_topic_info_dict = PyDict_New();
    if (!py_topic_info_dict) {
      goto fail;
    }
    // Populate keyword arguments
    // A success returns 0, and a failure returns -1
    int set_result = 0;
    set_result += PyDict_SetItemString(py_topic_info_dict, "qos_profile", py_qos_profile);
    set_result += PyDict_SetItemString(py_topic_info_dict, "node_name", py_node_name);
    set_result += PyDict_SetItemString(py_topic_info_dict, "node_namespace", py_node_namespace);
    set_result += PyDict_SetItemString(py_topic_info_dict, "topic_type", py_topic_type);
    set_result += PyDict_SetItemString(py_topic_info_dict, "gid", py_gid);
    if (set_result != 0) {
      goto fail;
    }
    return py_topic_info_dict;

fail:
    Py_XDECREF(py_qos_profile);
    Py_XDECREF(py_node_name);
    Py_XDECREF(py_node_namespace);
    Py_XDECREF(py_topic_type);
    Py_XDECREF(py_gid);
    Py_XDECREF(py_topic_info_dict);
    return NULL;
}
```

See https://docs.python.org/3/c-api/refcounting.html#c.Py_XDECREF.
The goto patter for error handling in C is much more clearer, collecting all error handling code in one place and avoiding duplication.